src/coreclr/jit/loopcloning.cpp

```diff
@@ -3190,6 +3268,9 @@ PhaseStatus Compiler::optCloneLoops()
         JITDUMP("Recompute reachability and dominators after loop cloning\n");
         // TODO: recompute the loop table, to include the slow loop path in the table?
         fgUpdateChangedFlowGraph(FlowGraphUpdates::COMPUTE_DOMS);
+
+        m_dfs   = fgComputeDfs();
+        m_loops = FlowGraphNaturalLoops::Find(m_dfs);
```

Comment:
The "TODO" above points out that we don't rebuild the loop table, but obviously here we do rebuild the new loop table.

Does anyone use this after this point?

If they used it and it wasn't recomputed, would it be incorrect, or just missing data?

It looks like ProfileSynthesis has its own copy of `FlowGraphDfsTree`, `FlowGraphNaturalLoops`. Should it just use the ones in `Compiler`?

Reply:
No, this is unused after this point. We are going to need to do this recomputation eventually, so I simply added it now to get a more proper view of the TP cost. 
I'll get rid of the TODO as part of removing the old loop stuff.

> If they used it and it wasn't recomputed, would it be incorrect, or just missing data?

If you change the flow graph and then recompute the DFS, then `BasicBlock::bbPostorderNum` will become invalid and mappings into `m_dfs` and `m_loops` will give nonsensical data. It's why I think it makes sense to assert that `m_dfs` matches the current flow graph in post phase checks (and null it out when we then no longer need it, i.e. after RBO). As long as the DFS isn't recomputed you can add and remove new blocks, and queries in the DFS tree/loop structures will just give incorrect (outdated) results for the affected blocks. For example, if you add a new block that is clearly reachable, `m_dfs->Contains(newBlock)` will return false incorrectly. In loop cloning we run over the loops parents before children which ensures we do not invalidate any of the stored data for any future loop we will see; thus we can keep acting on it. Loop unrolling (as I mentioned above) will be a bit harder.

> It looks like ProfileSynthesis has its own copy of FlowGraphDfsTree, FlowGraphNaturalLoops. Should it just use the ones in Compiler?

I don't think so. I think if we can avoid keeping ambient state like this then we should; otherwise we need to take care to invalidate the information in the right places.